src/Propulsion.MessageDb/ReaderCheckpoint.fs

```diff
@@ -0,0 +1,81 @@
+module Propulsion.MessageDb.ReaderCheckpoint
+
+open Npgsql
+open NpgsqlTypes
+open Propulsion.Feed
+open Propulsion.Infrastructure
+open System
+
+let createIfNotExists (conn : NpgsqlConnection, schema: string) =
+    let cmd = conn.CreateCommand()
+    cmd.CommandText <- $"
+      create table if not exists {schema}.propulsion_checkpoints (
+        stream_name text not null,
+        consumer_group text not null,
+        global_position bigint not null,
+        primary key (stream_name, consumer_group)
+      );
+    "
+    cmd.ExecuteNonQueryAsync() |> Async.AwaitTaskCorrect |> Async.Ignore<int>
+
+let commitPosition (conn : NpgsqlConnection, schema: string) (stream : string) (consumerGroup : string) (position : int64) = async {
+     let cmd = conn.CreateCommand()
+     cmd.CommandText <-
+         $"insert into {schema}.propulsion_checkpoints(stream_name, consumer_group, global_position)
+           values (@StreamName, @ConsumerGroup, @GlobalPosition)
+           on conflict (stream_name, consumer_group)
+           do update set global_position = @GlobalPosition;"
+     cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, stream) |> ignore
+     cmd.Parameters.AddWithValue("ConsumerGroup", NpgsqlDbType.Text, consumerGroup) |> ignore
+     cmd.Parameters.AddWithValue("GlobalPosition", NpgsqlDbType.Bigint, position) |> ignore
+
+     let! ct = Async.CancellationToken
+     do! cmd.ExecuteNonQueryAsync(ct) |> Async.AwaitTaskCorrect |> Async.Ignore<int> }
+
+let tryGetPosition (conn : NpgsqlConnection, schema : string) (stream : string) (consumerGroup : string) = async {
+    let cmd = conn.CreateCommand()
+    cmd.CommandText <-
+        $"select global_position from {schema}.propulsion_checkpoints where stream_name = @StreamName and consumer_group = @ConsumerGroup"
+
+    cmd.Parameters.AddWithValue("StreamName", NpgsqlDbType.Text, stream) |> ignore
+    cmd.Parameters.AddWithValue("ConsumerGroup", NpgsqlDbType.Text, consumerGroup) |> ignore
+
+    use reader = cmd.ExecuteReader()
+    let! ct = Async.CancellationToken
+    let! hasRow = reader.ReadAsync(ct) |> Async.AwaitTaskCorrect
+    return if hasRow then Some (reader.GetInt64 0) else None }
+
+type NpgsqlCheckpointStore(connString : string, schema: string, consumerGroupName, defaultCheckpointFrequency) =
+
+    let streamName source tranche =
+        match SourceId.toString source, TrancheId.toString tranche with
+        | s, null -> s
+        | s, tid -> String.Join("_", s, tid)
+
+    member _.CreateSchemaIfNotExists() = async {
+        use conn = new NpgsqlConnection(connString)
```

Comment:
most of this stuff should probably a `CT -> Task<NpgsqlConnection>`

with some helpers sitting in a `module Core` in `MessageDbSource`